Apromore-Custom-Plugins/Log-Importer-Portal/src/main/java/org/apromore/plugin/portal/logimporter/LogImporterController.java

```diff
@@ -398,23 +406,48 @@ protected LogModel getLogModel() throws Exception {
         return logModel;
     }
 
-    private void storeMappingAsJSON(LogMetaData logMetaData, Log log)
+    private void storeMetadataAsJSON(LogMetaData logMetaData, Log log)
             throws UserNotFoundException {
 
         String username = portalContext.getCurrentUser().getUsername();
-        String jsonStr = "";
+        String logMetadataJsonStr = "";
+        String perspectiveJsonStr = "";
 
         // Creating Object of ObjectMapper define in Jakson Api
         ObjectMapper objectMapper = new ObjectMapper();
         try {
-            jsonStr = objectMapper.writeValueAsString(logMetaData);
+            logMetadataJsonStr = objectMapper.writeValueAsString(logMetaData);
+            perspectiveJsonStr = objectMapper.writeValueAsString(getPerspectiveFromLogMetadata(logMetaData));
         } catch (IOException e) {
             LOGGER.error("Unable to convert log metadata into JSON; will store an empty string instead",
                     e);
         }
 
-        userMetadataService.saveUserMetadata("Default CSV schema mapping name", jsonStr,
+        userMetadataService.saveUserMetadata("Default CSV schema mapping name", logMetadataJsonStr,
                 UserMetadataTypeEnum.CSV_IMPORTER, username, log.getId());
+        userMetadataService.saveUserMetadata(UserMetadataTypeEnum.PERSPECTIVE_TAG.toString(), perspectiveJsonStr,
+                UserMetadataTypeEnum.PERSPECTIVE_TAG, username, log.getId());
+    }
+
+    public List<String> getPerspectiveFromLogMetadata(LogMetaData logMetaData) {
+
+        List<String> result = new ArrayList<>();
+        List<Integer> perspectivePos = logMetaData.getPerspectivePos();
+        List<String> headers = logMetaData.getHeader();
+
+        if (logMetaData.getActivityPos() != HEADER_ABSENT) {
+            result.add(XConceptExtension.KEY_NAME);
+        }
+
+        if (logMetaData.getResourcePos() != HEADER_ABSENT) {
+            result.add(XOrganizationalExtension.KEY_RESOURCE);
+        }
```

Comment:
What happens if the Activity column has actually a different attribute name in the log file from XConceptExtension.KEY_NAME ("concept:name")? For example, if the CSV file being imported doesn't have any attribute with name = "concept:name" but it has an attribute called "action" with the same meaning as "concept:name". If the above code is executed, the list of perspectives will contain "concept:name" rather than "action" which is wrong. Later, when a plugin retrieves this perspective, it won't be able to find any attribute with name = "concept:name" because the log has only attribute with name = "action.

The same for XOrganizationalExtension.KEY_RESOURCE.

Reply:
> As discussed, can we add unit tests for the Perspective data in LogMetaData? These are tests for LogImporterLogic.
> Suggest we do the following:
> 
> * Move getPerspectiveFromLogMetadata() from LogImporterController to LogMetaData as it should be (rename: getPerspectiveAttributes or getPerspectives)
> * Add simple LogMetaData objects for testing along with simple logs for different scenarios, e.g. default concept:name and org;resource prespectives, no default perspectives with a tagged activity attribute, no default/no activity attributes with custom event attribute as perspective.
> * Add tests for getPerspectiveFromLogMetadata
> * Add tests for LogProcessor.processLog regarding the Perspective data.

Unit tests added as per discussion.

Reply:
The conversion from attribute name to standard XES attribute identifier ("concept:name", "org:resource") is already handled inside `LogProcessor`. So no action is required here as per discussion.